numpy/_core/tests/test_stringdtype.py

```diff
@@ -1159,3 +1177,187 @@ def test_strip(string_array, unicode_array):
+        ])
+        self.s_empty = ""
+        self.s_short = "01234"
+        self.s_medium = "abcdefghijklmnopqrstuvwxyz"
+        self.s_long = "-=+" * 100
+        self.a = np.array(
+            [self.s_empty, self.s_short, self.s_medium, self.s_long],
+            self.dtype)
+
+    def get_view(self, a):
+        # Cannot view a StringDType as anything else directly, since
+        # it has references. So, we use a stride trick hack.
+        from numpy.lib._stride_tricks_impl import DummyArray
+        interface = dict(a.__array_interface__)
+        interface['descr'] = self.view_dtype.descr
+        interface['typestr'] = self.view_dtype.str
+        return np.asarray(DummyArray(interface, base=a))
+
+    def get_flags(self, a):
+        return self.get_view(a)['size_and_flags'] & 0xf0
+
+    def is_short(self, a):
+        return self.get_flags(a) == self.INITIALIZED | self.OUTSIDE_ARENA
+
+    def is_on_heap(self, a):
+        return self.get_flags(a) == (self.INITIALIZED
+                                     | self.OUTSIDE_ARENA
+                                     | self.LONG)
+
+    def is_missing(self, a):
+        return self.get_flags(a) & self.MISSING == self.MISSING
+
+    def in_arena(self, a):
+        return (self.get_flags(a) & (self.INITIALIZED | self.OUTSIDE_ARENA)
+                == self.INITIALIZED)
+
+    def test_setup(self):
+        is_short = self.is_short(self.a)
+        assert_array_equal(is_short, np.strings.str_len(self.a) <= 15)
+        assert_array_equal(self.in_arena(self.a), ~is_short)
+        assert_array_equal(self.is_on_heap(self.a), False)
+        assert_array_equal(self.is_missing(self.a), False)
+        view = self.get_view(self.a)
+        sizes = np.where(is_short, view['size_and_flags'] & 0xf,
+                         view['size'])
+        assert_array_equal(sizes, np.strings.str_len(self.a))
+        assert_array_equal(view['xsiz'][2:],
+                           np.void(b'\x00' * (self.sizeofstr // 4 - 1)))
+        # Check that the medium string uses only 1 byte for its length
+        # in the arena, while the long string takes 8 (or 4).
+        offsets = view['offset']
+        assert offsets[2] == 1
+        assert offsets[3] == 1 + len(self.s_medium) + self.sizeofstr // 2
+
+    def test_empty(self):
+        e = np.empty((3,), self.dtype)
+        assert_array_equal(self.get_flags(e), 0)
+        assert_array_equal(e, "")
+
+    def test_zeros(self):
+        z = np.zeros((2,), self.dtype)
+        assert_array_equal(self.get_flags(z), 0)
+        assert_array_equal(z, "")
+
+    def test_copy(self):
+        c = self.a.copy()
+        assert_array_equal(self.get_flags(c), self.get_flags(self.a))
+        assert_array_equal(c, self.a)
+
+    def test_arena_use_with_setting(self):
+        c = np.zeros_like(self.a)
+        assert_array_equal(self.get_flags(c), 0)
+        c[:] = self.a
+        assert_array_equal(self.get_flags(c), self.get_flags(self.a))
```

Comment:
neat, in the old implementation this would have required heap allocations because we'd be enlarging strings in an array that has already been initialized but now that we have the initialized flag this "just works".

Reply:
I was actually somewhat surprised it worked... But I agree it is good that it does.